src/Chapter3.hs

```diff
@@ -480,6 +522,54 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+type Wall = Bool
+
+type Church = Bool
+type Library = Bool
+data Building = Library | Church
+
+data Castle = Castle
+    {
+      __name__ :: String,
+      wall :: Wall
+    }
+
+newtype Houses = Houses
+    {
+      __people__ :: Int
+    }
+
+data MagicalCity = MagicalCity
+    {
+      castle :: Castle,
```

Comment:
Actually, this is optional, so the type could be:
```suggestion
      castle :: Maybe Castle,
```

Reply:
Oh, I see that you indecate castle availability by the empty/non-empty string. 
In this case,this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable